File: Google Docs (Document Writer)/Good Design/Java/DocumentWriter.java
import java.util.ArrayList;
import java.util.List;
import java.io.FileWriter;
import java.io.IOException;

interface Element {
    String render();
    String getElement();
}

class TextElement implements Element {
    private String text;

    public TextElement(String text) {
        this.text = text;
    }

    @Override
    public String render() {
        return "Text: " + text;
    }

    @Override
    public String getElement() {
        return text;
    }

}

class ImageElement implements Element {
    private String imageUrl;

    public ImageElement(String imageUrl) {
        this.imageUrl = imageUrl;
    }

    @Override
    public String render() {
        return "Image URL: [ " + imageUrl + " ]";
    }

    @Override
    public String getElement() {
        return imageUrl;
    }

}

interface DocumentWithText {
    void addText(String text);
}

interface DocumentWithImage {
    void addImage(String imageurl);
}

interface DocumentWithTextAndImage extends DocumentWithText, DocumentWithImage {
    List<Element> getElements();
}

class WordDocument {
    private List<Element> elements;

    public WordDocument() {
        this.elements = new ArrayList<>();
    }

    public void addText(String text) {
        elements.add(new TextElement(text));
    }

    public void addImage(String imageurl) {
        elements.add(new ImageElement(imageurl));
    }

    public List<Element> getElements() {
        return elements;
    }  
}


interface Renderer {
    String render(List<Element> elements);
}

class TextRenderer implements Renderer {
    @Override
    public String render(List<Element> elements) {
        StringBuilder sb = new StringBuilder();
        for (Element element : elements) {
            sb.append(element.render()).append("\n");
        }
        return sb.toString();
    }
}


interface Persistance {
    void save(String content, String name);
}

class SaveAsTextFile implements Persistance {
    @Override
    public void save(String content, String name) {
        try (FileWriter writer = new FileWriter(name)) {
            writer.write(content);
        } catch (IOException e) {
            e.printStackTrace();
        }
    }   
}

interface Formatter {
    Element format(Element Element);
}

interface TextFormatter extends Formatter{
}

class BoldTextFormatter implements TextFormatter{
    @Override
    public Element format(Element element){
        if(element instanceof TextElement){
            String originalText = ((TextElement) element).getElement();
            String boldText = "**" + originalText + "**";
            return new TextElement(boldText);
        }
        return element;
    }    
}

class ItalicTextFormatter implements TextFormatter{
    @Override
    public Element format(Element element){
        if(element instanceof TextElement){
            String originalText = ((TextElement) element).getElement();
            String italicText = "*" + originalText + "*";
            return new TextElement(italicText);
        }
        return element;
    }
}

public class DocumentWriter{
    public static void main(String[] args){
        
        WordDocument document = new WordDocument();
        document.addText("Hello, World!");
        document.addImage("http://example.com/image.png");
        document.addText("This is a sample document.");

        // Apply formatting
        TextFormatter boldFormatter = new BoldTextFormatter();
        TextFormatter italicFormatter = new ItalicTextFormatter();

        List<Element> formattedElements = new ArrayList<>();
        for (Element element : document.getElements()) {
            Element formattedElement = boldFormatter.format(element);
            formattedElement = italicFormatter.format(formattedElement);
            formattedElements.add(formattedElement);
        }

        // Render document
        Renderer renderer = new TextRenderer();
        String renderedContent = renderer.render(formattedElements);

        // Save document
        Persistance persistance = new SaveAsTextFile();
        persistance.save(renderedContent, "document.txt");

    }
}

File: Google Docs (Document Writer)/Good Design/Python/DocumentWriter.py
from typing import Protocol as protocols

class Element(protocols):
    def render(self) -> str : pass
    def getElements(self) -> str : pass

class TextElement(Element):
    def __init__(self, text: str):
        self.text = text

    def render(self) -> str:
        # do some modifications if needed
        return self.text

    def getElements(self) -> str:
        return self.text

class ImageElement(Element):
    def __init__(self, imagePath: str):
        self.imagePath = imagePath

    def render(self) -> str:
        # do some modifications if needed
        return f"<img src='{self.imagePath}' />"

    def getElements(self) -> str:
        return self.imagePath
    
class DocumentWithImage(protocols):
    def addImage(self, imageurl: str) -> None: pass

class DocumentWithText(protocols):
    def addText(self, text: str) -> None: pass

class Document(DocumentWithImage, DocumentWithText):
    def getElemets(self) -> list[Element] : pass

class TextDocument(Document):
    def __init__(self):
        self.elements: list[Element] = []

    def addText(self, text: str) -> None:
        text_element = TextElement(text)
        self.elements.append(text_element)

    def addImage(self, imageurl: str) -> None:
        image_element = ImageElement(imageurl)
        self.elements.append(image_element)

    def getElemets(self) -> list[Element]:
        return self.elements

class Persistence(protocols):
    def save(self, content:str, path:str) -> None : pass

class SaveAsTextFile(Persistence):
    def save(self, content: str, path: str) -> None:
        with open(path, 'w') as file:
            file.write(content)

class Renderer(protocols):
    def render(self, elements: list[Element]) -> str : pass

class TextRenderer(Renderer):
    def render(self, elements: list[Element]) -> str:
        rendered_content = ""
        for element in elements:
            rendered_content += element.render() + "\n"
        return rendered_content

class Formatter(protocols):
    def format(self, element: Element) -> Element : pass

class TextFormatter(Formatter): pass

class UpperCaseFormatter(TextFormatter):
    def format(self, element: Element) -> Element:
        if isinstance(element, TextElement):
            element.text = element.text.upper()
        return element

class LowerCaseFormatter(TextFormatter):
    def format(self, element: Element) -> Element:
        if isinstance(element, TextElement):
            element.text = element.text.lower()
        return element

class BoldFormatter(TextFormatter):
    def format(self, element: Element) -> Element:
        if isinstance(element, TextElement):
            element.text = f"**{element.text}**"
        return element

if __name__=="__main__":
    
    document = TextDocument()
    document.addText("Hello, World!")
    document.addImage("path/to/image.png")
    document.addText("This is a sample document.")

    # Apply formatters
    formatters = [UpperCaseFormatter(), BoldFormatter()]
    for i, element in enumerate(document.getElemets()):
        for formatter in formatters:
            element = formatter.format(element)
        document.getElemets()[i] = element

    # Render document
    renderer = TextRenderer()
    rendered_content = renderer.render(document.getElemets())

    print(rendered_content)

    # Save document
    persistence = SaveAsTextFile()
    persistence.save(rendered_content, "output.txt")
    print("Document saved as output.txt")
